### PR TITLE
ISSUE-130 Align server/client side apis with service api.

### PR DESCRIPTION
**Pull request by satishd**
- Align server/client side apis with service api.    
- Aligned without breaking the existing client APIs.

### CODE REVIEW

**Comment by satishd** [comment]
Closes #130

**Comment by harshach** [comment]
+1

File: schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java [hunk 1]
```diff
@@ -97,176 +90,51 @@
...
-
-
     /**
      * @param schemaName   name identifying a schema
      * @param toSchemaText text representing the schema to be checked for compatibility
-     * @return true if the given {@code toSchemaText} is compatible with the latest version of the schema with id as {@code schemaName}.
+     * @return true if the given {@code toSchemaText} is compatible with all the versions of the schema with name as {@code schemaName}.
      * @throws SchemaNotFoundException if there is no schema metadata registered with the given {@code schemaName}
      */
     boolean isCompatibleWithAllVersions(String schemaName, String toSchemaText) throws SchemaNotFoundException;
```
Comment (michaelandrepearce) [inline schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java]: Should this not be declared at a service level, should be a remove call.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/ISchemaRegistryService.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    Collection<SchemaProviderInfo> getSupportedSchemaProviders();
+
+    /**
+     * Registers information about a schema.
+     *
+     * @param schemaMetadata metadata about schema.
+     * @return id of the registered schema which is successfully registered now or earlier.
+     */
+    Long registerSchemaMetadata(SchemaMetadata schemaMetadata);
```
Comment (michaelandrepearce) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/ISchemaRegistryService.java]: for the small breaking change should we not rename this as add, or rename the adds as register (seems inconsistent)

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java [hunk 3]
```diff
@@ -18,7 +18,9 @@
 /**
  * Indicates that schema type is invalid.
  */
-public class UnsupportedSchemaTypeException extends Exception {
+public class UnsupportedSchemaTypeException extends RuntimeException {
```
Comment (michaelandrepearce) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java]: Do we really need to be making RuntimeExceptions, its a lot harder for end users to handle as they're easily left off interfaces etc. We should be avoiding IMO runtime exceptions

**Comment by michaelandrepearce** [comment]
@satishd thanks, have left some comments, cheers.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java (hunk 3 above)
Comment (satishd) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java]: This Exception is not really exposed on client side. Extending RTE is a controversial topic and it has both advantages/disadvantages. My preference is to avid checked exceptions.

File: schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java (hunk 1 above)
Comment (satishd) [inline schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java]: There is a different method signature in `ISchemaRegistry` and that is the reason why this was not moved to `ISchemaRegistryService`. Ideally, I would have deprecated this method and introduced the existing one in `ISchemaRegistry`. I was not sure whether I would add that now, let me know your opinion.

**Comment by satishd** [comment]
@michaelandrepearce Thanks for reviewing, updated with my comments.

File: schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java (hunk 1 above)
Comment (michaelandrepearce) [inline schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java]: I think if the effort is to try align then where we can align be deprecating and old and adding a new it's worth it, as long as there's a pre agreed time the deprecated would be removed eg would revomed in next major +1 release

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/ISchemaRegistryService.java (hunk 2 above)
Comment (michaelandrepearce) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/ISchemaRegistryService.java]: Fair enough. Maybe as per the other comment, add the aligned name and keep the deprecated one with agreed removal time scales

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java (hunk 3 above)
Comment (michaelandrepearce) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/errors/UnsupportedSchemaTypeException.java]: Fair enough

File: schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java (hunk 1 above)
Comment (satishd) [inline schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java]: @harshach I plan to change couple of these APIs as deprecated and remain till next release. Let me know your opinion.

Comment (harshach) [inline schema-registry/client/src/main/java/com/hortonworks/registries/schemaregistry/client/ISchemaRegistryClient.java]: @satishd  sounds good to me

**Comment by satishd** [comment]
@michaelandrepearce @harshach Updated with deprecated API changes.

**Comment by harshach** [comment]
still +1 on merging this.

**Pull request merged**